### ISSUE TITLE
Deferring EOI until after RestoreTPL() causes system time to stop (Bugzilla Bug 4162)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9256]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 4162
Date: 2022-11-21T10:47:36+00:00
From: Michael Brown &lt;<mcb30>&gt;
To:   unassigned &lt;<unassigned>&gt;
CC:   @ajfish, igor.druzhinin, @lersek, @mdkinney, pbonzini

See also:     https://bugzilla.tianocore.org/show_bug.cgi?id=2815
Last updated: 2022-12-23T12:00:59+00:00

**Comment by tianocore-issues** [comment]
## Comment 20070
Date: 2022-11-21 10:47:36 +0000
From: Michael Brown &lt;<mcb30>&gt;

- Industry Specification: ---
- Release Observed: ['EDK II Master', 'edk2-stable202008', 'edk2-stable202011', 'edk2-stable202102', 'edk2-stable202105']
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): unassigned &lt;<unassigned>&gt;

Commit 239b50a86 (&quot;OvmfPkg: End timer interrupt later to avoid stack overflow under load&quot;) deferred the call to SendApicEoi() until after the call to RestoreTPL(), in order to avoid the potential stack underflow as described in bug 2815.

This has the unwanted side effect that any callbacks invoked as a result of RestoreTPL() will now run before the EOI is sent to the APIC.  Until the callbacks return, no further timer interrupts will be raised and the system time value in mEfiSystemTime will remain unchanging.

If any callbacks invoked by RestoreTPL() include a timeout loop based on mEfiSystemTime (e.g. via SetTimer() and CheckEvent()), this loop will never time out and the system will appear to hang.

This situation can be reproduced by e.g. attaching a USB network card connected via an xHCI USB controller.  The &quot;system poll timer&quot; callback to MnpSystemPoll() will eventually cause the following sequence of events:

- Hardware timer interrupt occurs
- TimerInterruptHandler() in LocalApicTimerDxe is called
- TimerInterruptHandler() calls RaiseTPL(TPL_HIGH_LEVEL)
- TimerInterruptHandler() calls RestoreTPL() before sending EOI
- RestoreTPL() triggers a call to MnpSystemPoll()
- MnpSystemPoll() ends up calling through to XhcExecTransfer() to poll the USB device
- XhcExecTransfer() enters a loop waiting for the transfer to complete or time out
- XhcExecTransfer() uses CheckEvent() to check for the timeout condition
- Since EOI has not been sent, no further timer interrupts will occur and so CheckEvent() always returns EFI_NOT_READY and XhcExecTransfer() waits indefinitely in its timeout loop
- User sees a totally stuck system that is unresponsive to any input other than a hard reset


This problem does not affect EHCI or UHCI controllers since they do not rely on SetTimer()/CheckEvent() to check for timeout.


The code pattern in LocalApicTimerDxe&apos;s TimerInterruptHandler() looks highly dubious: it is surprising to see an interrupt handler itself enabling and disabling interrupts.  I am unclear what the proper fix for this ought to be.

I&apos;m happy to code up a fix if someone can tell me what the correct approach should be.

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20071
Date: 2022-11-22 05:17:36 +0000
From: @lersek

I&apos;ve been staring at this for hours now. Here&apos;s what I can say.

I think I was wrong to ACK commit 239b50a86 (the patch for bug#2815).
That commit effectively disables interrupt nesting. (The patch does not
interfere with interrupt delivery, but it prevents interrupt generation,
by delaying EOI.)

However, interrupt nesting -- as I now think I understand -- is a
necessity for edk2

(1) UEFI drivers and a running UEFI application, taken all together, do
    not follow the &quot;one central main loop + non-interruptible, quick ISR
    + bottom halves&quot; style. The drivers&apos; timer notifications must be
    processed orthogonally to the application being aware or not. This
    is only possible if TPL_CALLBACK notification functions of the
    drivers actually interrupt the TPL_APPLICATION code of the
    application. For that, the TPL_CALLBACK notification functions, such
    as MnpSystemPoll() --&gt; XhcExecTransfer(), have to run in interrupt
    handler context.

(2) In turn, TPL_CALLBACK notification functions (running in interrupt
    handler context, as stated above), such as XhcExecTransfer(), are
    permitted to call the CreateEvent(), SetTimer() and CheckEvent()
    boot services, per UEFI spec / TPL restrictions.

(3) In turn, those boot services rely on timer interrupts being
    generated and delivered while they are being used, for example in a
    polling loop, such as the one seen in XhcExecTransfer().

Ergo, timer interrupts must be generated and delivered while other stuff
is happening in interrupt handler context (such as a TPL_CALLBACK
notification running).

This statement happens to be the definition of interrupt nesting -- the
handler itself ought to be, ultimately, interruptible and reentrant.

Specifically, MnpSystemPoll() is queued at TPL_CALLBACK, hence its
callee XhcExecTransfer() too executes at TPL_CALLBACK -- still in
interrupt handler context (there is no other context to run
MnpSystemPoll() in). At TPL_CALLBACK, XhcExecTransfer() is permitted to
use CreateEvent(), SetTimer() and CheckEvent(). CheckEvent() will only
work if timer interrupts are being generated and delivered while
XhcExecTransfer() is looping.

Therefore we should revert commit 239b50a86, and fix the interrupt storm
(or interrupt burst) described in bug#2815 differently.

Here&apos;s how the interrupt storm / burst mis-behaves, as far as I can
guesstimate. Namely, with commit 239b50a86 *reverted*:

(a) When the first interrupt of the delayed burst is delivered, we most
    likely interrupt TPL_APPLICATION code. Therefore in
    TimerInterruptHandler() [OvmfPkg/8254TimerDxe/Timer.c], we call
    &quot;CoreRaiseTpl (TPL_HIGH_LEVEL)&quot; against TPL_APPLICATION.

(b) This manually disables interrupt delivery (somewhat needlessly I
    think, because we&apos;re already in interrupt context). It also sets
    &quot;gEfiCurrentTpl&quot; to TPL_HIGH_LEVEL, and returns TPL_APPLICATION.

(c) We signal EOI -- interrupts can now be generated again, but not
    delivered just yet.

(d) There is likely a timer event notification function that we queue.
    (Said function is probably &quot;way overdue&quot;, because the timer
    interrupts in question have been significantly delayed in the first
    place, due to Xen descheduling the vCPU for a long time.)

(e) TimerInterruptHandler() calls CoreRestoreTpl() with
    NewTpl=TPL_APPLICATION.

(f) In CoreRestoreTpl(), we see that there is at least one TPL_CALLBACK
    function queued (from step (d)), so we set &quot;PendingTpl&quot; to
    TPL_CALLBACK. Because NewTpl (TPL_APPLICATION) is less than
    PendingTpl (TPL_CALLBACK), we proceed to dispatching the function as
    follows.

(g) We set &quot;gEfiCurrentTpl&quot; to TPL_CALLBACK.

(h) Because interrupt nesting is necessary for the sake of boot services
    working inside callbacks, we reenable interrupt delivery, in
    preparation for calling the the queued notification function.

(i) The second interrupt of the burst is delivered immediately, before
    we can advance to calling the notification function.

(j) We enter TimerInterruptHandler() asynchronously, on a new stack
    frame.

(k) From TimerInterruptHandler(), we call CoreRaiseTpl (TPL_HIGH_LEVEL)
    again.

(l) This disables interrupt delivery (needlessly, similarly to step
    (b)), sets &quot;gEfiCurrentTpl&quot; to TPL_HIGH_LEVEL, and returns
    TPL_CALLBACK (!!!). Note the return value difference with step (b).

(m) We signal EOI.

(n) The timer event notification function, which we failed to actually
    dispatch after step (h), is still pending (still overdue), therefore
    we queue it again (or at least we &quot;mark it&quot; as pending,
    idempotently; doesn&apos;t matter).

(o) TimerInterruptHandler() calls CoreRestoreTpl() with
    NewTpl=TPL_CALLBACK.

(p) In CoreRestoreTpl(), while we see that there is still at least one
    TPL_CALLBACK function queued (and so we set &quot;PendingTpl&quot; to
    TPL_CALLBACK again), NewTpl is now TPL_CALLBACK, therefore we break
    out of the loop that handles the different task priority level
    queues, don&apos;t reenable interrupt delivery, and don&apos;t try to dispatch
    anything. This seems correct!

(q) We set &quot;gEfiCurrentTpl&quot; to TPL_CALLBACK.

(r) At the end of CoreRestoreTpl(), we reenable interrupt delivery --
    and the third interrupt of the burst enters TimerInterruptHandler()
    on a new frame.

    I think this is the problem!

    This is what commit 239b50a86 attempts to defuse, by keeping
    interrupt *generation* disabled, by delaying the EOI. After
    CoreRestoreTpl() returns, commit 239b50a86 re-disables interrupt
    delivery -- effectively undoing step (r) under the protection of the
    *withheld* EOI --, and then commit 239b50a86 finally signals the
    EOI. Interrupt delivery is then reenabled by IRET (refer to
    &lt;https://bugzilla.tianocore.org/show_bug.cgi?id=2815#c7&gt;).

    Instead, CoreRestoreTpl() should just *not* re-enable interrupt
    delivery here. It should only return to TimerInterruptHandler(), and
    allow TimerInterruptHandler() to complete. When that happened, we&apos;d
    execute an IRET, and that would reenable interrupt delivery
    simultaneously with popping the frame off the stack. We&apos;d
    subsequently return to just after step (h), with interrupt delivery
    enabled. That&apos;s where the entire burst would be coalesced (repeat
    the sequence from (i) through (q), many times), and after the final
    interrupt of the burst, proceed to dispatching the queued
    notification function.

So here&apos;s my opinion / proposal:

(i) In CoreRaiseTpl(), when the TPL is raised from below HIGH to HIGH,
    and we&apos;re about to disable interrupts, we should check *and record*
    whether we&apos;re already running in an interrupt handler context. Put
    differently, if interrupt delivery is already disabled, then don&apos;t
    try to disable it again; instead, just remember the fact, in a new
    global variable.

(ii) In CoreRestoreTpl(), the CoreDispatchEventNotifies() must be
     correctly *bracketed* by unmasking and re-masking interrupt
     delivery. The unmasking is already there, but the re-masking is
     missing. If we enter the loop with interrupt delivery turned off,
     we must exit the loop the same way.

(iii) At the very end of CoreRestoreTpl(), we must only unmask
      (re-enable) interrupt delivery if we&apos;re not already in interrupt
      handler context. That is, if we recorded, in step (i), that
      interrupts had been disabled before the TPL was going to be raised
      to HIGH, then do not reenable interrupt delivery here either, when
      lowering the TPL from HIGH.

gCpu-&gt;GetInterruptState() can be used to determine if interrupt delivery
is already disabled.

**Comment by tianocore-issues** [comment]
## Comment 20073
Date: 2022-11-22 05:42:25 +0000
From: @lersek

(Small, tangential correction: commit 239b50a86370 does interfere with both interrupt generation and interrupt delivery. The goal of the commit is to *undo* the final interrupt delivery enablement in RestoreTPL. The commit works as follows: first, EOI is delayed until after RestoreTPL, therefore, even though RestoreTPL still enables delivery at the end, there is nothing to deliver because nothing is generated by the interrupt controller. Second, after RestoreTPL returns, interrupt delivery is immediately disabled. Third, the EOI is sent. This ordering of steps #2 and #3 ensures that there is no window when an interrupt could be generated, but would not be masked, and hence be delivered.

The regression is that delaying the EOI until after RestoreTPL kills timer interrupts for notification functions called from within RestoreTPL, even though those notification functions depend on timer interrupts, whenever they call CheckEvent(), for example.

Therefore the right approach is not to &quot;undo&quot; RestoreTPL&apos;s final interrupt delivery enablement, from TimerInterruptHandler() -- instead, modify RestoreTPL *itself* so that the delivery reenablement step in question only be done in case we&apos;re not already in interrupt context. If we&apos;re already in interrupt context, then the IRET -- which follows upon TimerInterruptHandler() returning -- will reenable interrupt delivery *without* the risk of recursively entering TimerInterruptHandler().)

**Comment by tianocore-issues** [comment]
## Comment 20074
Date: 2022-11-22 05:55:57 +0000
From: @lersek

FWIW this reminds me slightly of spin_lock_irqsave() in the Linux kernel:

https://www.kernel.org/doc/html/v5.0/kernel-hacking/locking.html#hard-irq-context

&gt;| spin_lock_irqsave() (include/linux/spinlock.h) is a variant which saves
&gt;| whether interrupts were on or off in a flags word, which is passed to
&gt;| spin_unlock_irqrestore(). This means that the same code can be used inside
&gt;| an hard irq handler (where interrupts are already off) and in softirqs
&gt;| (where the irq disabling is required).

IMO that&apos;s what we want from the RaiseTPL and RestoreTPL boot services. They should work both when normal driver code raises the TPL to HIGH (and therefore explicit interrupt disabling is needed), and when the low-level timer interrupt handler raises the TPL to HIGH in interrupt context (and therefore interrupt delivery is managed externally to RaiseTPL/RestoreTPL).

**Comment by tianocore-issues** [comment]
## Comment 20075
Date: 2022-11-22 10:50:15 +0000
From: Michael Brown &lt;<mcb30>&gt;

Laszlo: thank you very much for the extremely detailed analysis!  I agree completely that it does feel very similar to spin_lock_irqsave().  Since we can&apos;t viably change the API for RaiseTPL()/RestoreTPL(), I&apos;m trying to think through the potential implications of holding the necessary flags variable in a single global, to convince myself that it&apos;s provably safe to do so.

Michael

**Comment by tianocore-issues** [comment]
## Comment 20076
Date: 2022-11-22 11:50:50 +0000
From: Michael Brown &lt;<mcb30>&gt;

I _think_ that a global flag (&quot;mRaisedFromIsr&quot;) cannot work, given the following possible sequence of events:

- Hardware interrupt occurs

- ISR calls RaiseTPL() to raise to TPL_HIGH_LEVEL.  RaiseTPL() records the global flag mRaisedFromIsr=TRUE based on gCpu-&gt;GetInterruptState().

- ISR calls RestoreTPL() to restore back to TPL_APPLICATION.  As part of this a callback is invoked at TPL_CALLBACK.

- The callback chooses (for some reason) to call RaiseTPL() to raise to TPL_HIGH_LEVEL.  RaiseRPL() overwrites the global flag mRaisedFromIsr=FALSE, since it has been called with interrupts (temporarily) enabled.

At this point, all record of the original mRaisedFromIsr=TRUE state has been lost.

Using a global counter instead of a boolean has similar problems: we don&apos;t have the ability to correctly pair the interrupt-restoring behaviour of each RestoreTPL() with its originating RaiseTPL() invocation.

Changing the return value from RaiseTPL() to include an &quot;interrupts-enabled&quot; bit would work but would be a breaking API change and so is not permissible.

Trying to think of a viable approach...

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20082
Date: 2022-11-23 12:10:38 +0000
From: Michael Brown &lt;<mcb30>&gt;

My current thinking is this:

- At the point that TimerInterruptHandler() calls RestoreTPL(), it records in a static variable (e.g. &quot;RestoringTPL&quot;) the TPL that it is attempting to restore back down to.

- When a nested interrupt occurs, this gives us a way for the new (inner) interrupt to know the TPL that the old (outer) interrupt was attempting to restore back down to.

- At the start of TimerInterruptHandler(), we compare OriginalTPL against RestoringTPL.

- If these are equal, then we know that the inner interrupt occurred during the outer interrupt handler&apos;s call to RestoreTPL() but _after_ all notifications had been dispatched by the outer interrupt&apos;s call to RestoreTPL().  This is the situation that arises from step (r) in Laszlo&apos;s breakdown of events, and the situation that we could prevent if RaiseTPL()/RestoreTPL() allowed for an extra parameter to handle the saved interrupts-enabled flag.

So, I think this allows us to at least cleanly detect the situation that we&apos;re trying to avoid.  What I&apos;m not sure about is how to then deal with it.  We could choose to deliberately defer the EOI for the inner interrupt until after the inner interrupt&apos;s call to RestoreTPL().  But I&apos;m not yet convinced that this works.

Michael

**Comment by tianocore-issues** [comment]
## Comment 20084
Date: 2022-11-23 14:19:19 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

This is a mess and is all because edk2 tries to mimic Windows IRQL without using the APIC&apos;s interrupt priority mechanism.

The problem is that there are _two_ interrupts enabled flags, namely the processor IF flag and gEfiCurrentTpl, and RaiseTPL only considers the latter.

In other words, there is an implicit &quot;RaiseTPL (TPL_HIGH_LEVEL)&quot; just because you&apos;re running with IF=0, and the problem that you want to do the &quot;RestoreTPL&quot; work after IRET, at IF=1.

On Windows, the interrupt handler would be able immediately set IF=1, because it can rely on the APIC&apos;s TPR to block further lower-priority interrupts.  And there would be no need for an explicit RaiseTPL(), because the current TPL is just the APIC&apos;s TPR rather than a global variable.

Also, on Windows raising an event would send an IPI to self (with a vector such that the TPR becomes either TPL_CALLBACK or TPL_NOTIFY).  The IPI would be processed after the EOI lowers the TPR, like this:

   /* TPR is high, preventing recursive interrupts */
   STI
   ...
   send IPI
   ...
   CLI
   send EOI
   /* TPR is low, but IF=0 so IPI is delayed until after IRET */
   IRET

So the Disable interrupts/EOI sequence does make sense, but RestoreTPL must somehow delay the event work until after the IRET in order to avoid exploding the stack.

Well, pre-APIC Windows NT did it so it must be possible.  But I think it requires some black magic in the IDT handlers.

For sure RaiseTPL() must be changed to ignore gEfiCurrentTpl and return TPL_HIGH_LEVEL if IF=0.  Then the RaiseTPL/RestoreTPL pair in TimerInterruptHandler becomes a glorified nop, and there is no need to disable the interrupts in TimerInterruptHandler before sending EOI.

To ensure the events are processed after interrupts are re-enabled, you could patch the return address of the IRET.  That is, the sequence would be

   /* in IDT handler */
   save all registers
   call handler
     /* IF=0 is high, preventing recursive interrupts */
     ...
     CoreNotifyEvent()
     ...
     send EOI

   /* back to IDT handler */
   oldTpl = RaiseTpl(TPL_HIGH_LEVEL);
   if any events for TPL &lt;= oldTpl
     mangle stack to return to EventProcessingTrampoline after IRET
   else
     RestoreTpl(oldTpl);

   restore all registers
   IRET

EventProcessingTrampoline:
   save all registers
   fetch oldTpl from stack frame
   RestoreTPL(oldTpl)
   restore all registers
   RET                    ; RET 4 on 32-bit, RET 8 on 64-bit

The stack mangling could be something like:
- pick the SS:RSP from the IRET frame
- push oldTpl onto that SS:RSP
- push the RIP from the IRET frame onto that SS:RSP
- change the RIP in the IRET frame to EventProcessingTrampoline

With some care because SS:RSP could point onto the very stack you&apos;re running on, i.e. something like (don&apos;t quote me on the order of the five fields):

  SS             oldTpl
  RSP            RIP
  RFLAGS    =&gt;   SS
  CS             RSP
  RIP            RFLAGS
                 CS
                 EventProcessingTrampoline


IIRC it&apos;s more complicated on 32-bit though because SS:SP is not always pushed. But maybe edk2 is always ring 0? If so it&apos;s not that big of a complication, as it would be just

  EFLAGS         oldTpl
  CS      =&gt;     EIP
  EIP            EFLAGS
                 CS
                 EventProcessingTrampoline

**Comment by tianocore-issues** [comment]
## Comment 20085
Date: 2022-11-23 15:23:20 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

&gt; For sure RaiseTPL() must be changed to ignore gEfiCurrentTpl and return
&gt; TPL_HIGH_LEVEL if IF=0.

This is not even necessary if RaiseTPL() is moved in the IDT handler.  Then all that&apos;s needed is an HasEventForTPL() function:


   /* in IDT handler */
   save all registers
   oldTpl = RaiseTpl(TPL_HIGH_LEVEL);
   call handler
     send EOI

   /* back to IDT handler with IF=0 */
   if HasEventsForTpl(oldTpl)
     mangle stack to return to EventProcessingTrampoline after IRET
   else
     RestoreTpl(oldTpl);

   restore all registers
   IRET

**Comment by tianocore-issues** [comment]
## Comment 20086
Date: 2022-11-23 15:41:16 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

Hmm, not entirely: RestoreTpl must not enable interrupts even if oldTpl &lt; TPL_HIGH_LEVEL. So there has to be a &quot;PrepareRestoreTplFromInterruptHandler&quot; function exported by MdeModulePkg/Core/Dxe/Event/Tpl.c, like

/* Highest interrupts-enabled TPL.  */
#define TPL_AFTER_INTERRUPT_LEVEL (TPL_HIGH_LEVEL - 1)

BOOL
EFIAPI
CorePrepareRestoreTplFromInterruptHandler (
  IN EFI_TPL NewTpl
  )
{
  EFI_TPL     OldTpl;
  EFI_TPL     PendingTpl;

  OldTpl = gEfiCurrentTpl;
  ASSERT(OldTpl == TPL_HIGH_LEVEL);
  if (NewTpl &gt; OldTpl) {
    DEBUG ((EFI_D_ERROR, &quot;FATAL ERROR - RestoreTpl from interrupt handler with NewTpl(0x%x) &gt; OldTpl(0x%x)\n&quot;, NewTpl, OldTpl));
    ASSERT (FALSE);
  }
  ASSERT (VALID_TPL (NewTpl));

  if (NewTpl == TPL_HIGH_LEVEL) {
    /* Cannot have any events.  */
    return FALSE;
  }

  if (gEventPending != 0) {
    PendingTpl = (UINTN) HighBitSet64 (gEventPending);
    if (PendingTpl &gt; NewTpl) {
      /* Keep a pretty high TPL until events are processed.  Interrupt state will
         be restored by caller.  */
      gEfiCurrentTpl = TPL_AFTER_INTERRUPT_LEVEL;
      return TRUE;
    }
  }
  gEfiCurrentTpl = NewTpl;
  /* Interrupt state will be restored by caller */
  return FALSE;
}

**Comment by tianocore-issues** [comment]
## Comment 20087
Date: 2022-11-23 15:53:37 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

As to how to implement it: a new protocol (gEfiInterruptEventProtocolGuid?) containing nothing but the new function could be exposed by DxeMain.inf.

**Comment by tianocore-issues** [comment]
## Comment 20088
Date: 2022-11-23 16:03:30 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

Okay, this should be the last update:

&gt; Hmm, not entirely: RestoreTpl must not enable interrupts even if oldTpl &lt; 
&gt; TPL_HIGH_LEVEL. So there has to be a &quot;PrepareRestoreTplFromInterruptHandler&quot; 
&gt; function exported by MdeModulePkg/Core/Dxe/Event/Tpl.c, like

Better: RestoreTpl must enable interrupts if oldTpl &lt; TPL_HIGH_LEVEL, but the interrupt handler must not.  So


   /* back to IDT handler with IF=0 */
   if HasEventsForTpl(oldTpl)
     mangle stack to return to EventProcessingTrampoline after IRET
   else
     RestoreTpl(oldTpl);

must become

   /* back to IDT handler with IF=0 */
   if (PrepareRestoreTplFromInterruptHandler(oldTpl))
     /* TRUE = Tpl is still high (TPL_AFTER_INTERRUPT_LEVEL) because there
        are events to process.  Interrupts will be enabled by IRET.  */
     mangle stack to return to EventProcessingTrampoline after IRET
   else
     /* FALSE = Tpl has been restored, interrupts will be enabled by IRET.  */
     ;

(followed by restoring all registers and IRET).

**Comment by tianocore-issues** [comment]
## Comment 20091
Date: 2022-11-24 07:28:37 +0000
From: Michael Brown &lt;<mcb30>&gt;

Paolo: thank you for also for such a detailed response.

I am not sure that I completely understand how this approach solves the problem.  In particular:

- Anything that would mimic the APIC interrupt priority mechanism cannot work.  UEFI design requires that the callbacks triggered from a timer interrupt are themselves allowed to wait on events triggered by further timer interrupts (i.e. by further instances of the exact same interrupt).

- Completing the event processing after IRET reduces the stack consumption per nested interrupt, but still leaves it greater than zero.  The stack exhaustion problem would be alleviated rather than solved.

Michael

**Comment by tianocore-issues** [comment]
## Comment 20092
Date: 2022-11-24 11:05:41 +0000
From: Michael Brown &lt;<mcb30>&gt;

Here is my proposed fix (after reverting 239b50a86):


--- a/OvmfPkg/LocalApicTimerDxe/LocalApicTimerDxe.c
+++ b/OvmfPkg/LocalApicTimerDxe/LocalApicTimerDxe.c
@@ -58,7 +58,10 @@ TimerInterruptHandler (
   IN EFI_SYSTEM_CONTEXT  SystemContext
   )
 {
-  EFI_TPL  OriginalTPL;
+  STATIC UINTN    RestoreTPLPending = 0;
+  STATIC EFI_TPL  TargetTPL = 0;
+  EFI_TPL         OldTargetTPL;
+  EFI_TPL         OriginalTPL;
 
   OriginalTPL = gBS-&gt;RaiseTPL (TPL_HIGH_LEVEL);
 
@@ -71,7 +74,51 @@ TimerInterruptHandler (
     mTimerNotifyFunction (mTimerPeriod);
   }
 
-  gBS-&gt;RestoreTPL (OriginalTPL);
+  //
+  // This is unfortunately messy.  UEFI requires us to support nested
+  // timer interrupts, but provides no way for RestoreTPL() to avoid
+  // re-enabling interrupts.  In a virtual machine, it is possible for
+  // a large burst of timer interrupts to arrive.  We must prevent
+  // such a burst from leading to stack underrun, while continuing to
+  // allow nested timer interrupts to occur.
+  //
+  // Since nested interrupts are permitted, we may be invoked as an
+  // inner interrupt handler while the outer interrupt handler is
+  // still inside its call to RestoreTPL().
+  //
+  // If the TPL at which we were invoked is higher than the target TPL
+  // for the outer handler&apos;s call to RestoreTPL(), then the outer
+  // handler&apos;s call to RestoreTPL() is still dispatching event
+  // notifications.  In this circumstance we must call RestoreTPL() to
+  // allow more event notifications to be dispatched, since those
+  // callbacks may themselves be waiting upon events.  We cannot avoid
+  // creating a new stack frame for this inner call to RestoreTPL(),
+  // but the total number of such stack frames is intrinsically
+  // limited by the number of distinct TPLs.
+  //
+  // If the TPL at which we were invoked is equal to the target TPL
+  // for the outer handler&apos;s call to RestoreTPL(), then the outer
+  // handler&apos;s call to RestoreTPL() must have finished dispatching all
+  // event notifications, and the interrupt therefore occurred at the
+  // point that RestoreTPL() was about to return to the outer handler.
+  // In this circumstance, we must not call RestoreTPL() here since
+  // doing so could lead to stack exhaustion.  Instead, we return from
+  // the ISR at TPL_HIGH_LEVEL and set a flag to trigger the outer
+  // handler into performing the deferred call to RestoreTPL().
+  //
+  ASSERT (OriginalTPL &gt;= TargetTPL);
+  if (OriginalTPL == TargetTPL) {
+    RestoreTPLPending |= (UINTN)(1 &lt;&lt; OriginalTPL);
+  } else {
+    OldTargetTPL = TargetTPL;
+    TargetTPL = OriginalTPL;
+    do {
+      RestoreTPLPending &amp;= ~(UINTN)(1 &lt;&lt; OriginalTPL);
+      gBS-&gt;RestoreTPL (OriginalTPL);
+      DisableInterrupts ();
+    } while (RestoreTPLPending &amp; (UINTN)(1 &lt;&lt; OriginalTPL));
+    TargetTPL = OldTargetTPL;
+  }

**Comment by tianocore-issues** [comment]
## Comment 20093
Date: 2022-11-24 11:26:51 +0000
From: Michael Brown &lt;<mcb30>&gt;

I have tested this suggested patch by simulating an interrupt storm (via calling gBS-&gt;Stall() inside TimerInterruptHandler()), and it does correctly limit the total stack consumption.

As far as I can tell, it is safe for us to defer the call to RestoreTPL() under the conditions described in the patch.  There will be a short period of time during which we are executing with the TPL at TPL_HIGH_LEVEL but with interrupts enabled, but this period can logically cover only the return from the outer RestoreTPL() to TimerInterruptHandler().  During this brief window, the only thing that could possibly interrupt the flow of execution would be another timer interrupt, which will gracefully do (almost) nothing when invoked from TPL_HIGH_LEVEL anyway.

@Laszlo @Paolo: any comments on my suggested fix?

Michael

**Comment by tianocore-issues** [comment]
## Comment 20094
Date: 2022-11-24 12:40:58 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

&gt; - Anything that would mimic the APIC interrupt priority mechanism cannot work.  UEFI 
&gt; design requires that the callbacks triggered from a timer interrupt are themselves 
&gt; allowed to wait on events triggered by further timer interrupts (i.e. by further 
&gt; instances of the exact same interrupt).

That should be okay because callbacks are invoked from their TPL.  Remember that the event processing loop is nothing but a fancy split of RestoreTpl(oldTpl) in two phases - first enabling interrupts and setting Tpl to TPL_HIGH_LEVEL-1, and then after IRET dispatching the pending events all the way down to oldTpl.  All that matters is the correct sequence of CLI, then EOI, then IRET (enabling further interrupts), then dispatch.

&gt; - Completing the event processing after IRET reduces the stack consumption per nested 
&gt; interrupt, but still leaves it greater than zero.  The stack exhaustion problem would be 
&gt; alleviated rather than solved.

That should also be fine because the each further instance of the event execution code would have a higher TPL than the previous one.  The maximum nesting would be

  interrupted code
    execute events with oldTpl = TPL_CALLBACK
      execute events with oldTpl = TPL_NOTIFY
        execute events with oldTpl = TPL_HIGH_LEVEL - 1

Since there are no events with a high-enough TPL, the inner TimerInterruptHandler would not trigger another execute-events code.

That said, I will look at your patch.

**Comment by tianocore-issues** [comment]
## Comment 20095
Date: 2022-11-24 13:33:06 +0000
From: Michael Brown &lt;<mcb30>&gt;

(In reply to Paolo Bonzini from comment #15)
&gt; That should be okay because callbacks are invoked from their TPL.  Remember
&gt; that the event processing loop is nothing but a fancy split of
&gt; RestoreTpl(oldTpl) in two phases - first enabling interrupts and setting Tpl
&gt; to TPL_HIGH_LEVEL-1, and then after IRET dispatching the pending events all
&gt; the way down to oldTpl.  All that matters is the correct sequence of CLI,
&gt; then EOI, then IRET (enabling further interrupts), then dispatch.

Other than some rearrangement of the stack contents, I think that gives the same sequence of events as EOI, STI (enable interrupts), then dispatch, then IRET (which will harmlessly enable the already enabled interrupts)?  Which I think is essentially the same as the pre-239b50a86 behaviour.

&gt; That should also be fine because the each further instance of the event
&gt; execution code would have a higher TPL than the previous one.  The maximum
&gt; nesting would be
&gt; 
&gt;   interrupted code
&gt;     execute events with oldTpl = TPL_CALLBACK
&gt;       execute events with oldTpl = TPL_NOTIFY
&gt;         execute events with oldTpl = TPL_HIGH_LEVEL - 1
&gt; 
&gt; Since there are no events with a high-enough TPL, the inner
&gt; TimerInterruptHandler would not trigger another execute-events code.

I think I must be misunderstanding something, sorry.

If the event processing is rearranged to come after the IRET, then the stack must have been rearranged to contain more information than was present when the interrupt occurred.  (It needs to contain at least the OriginalTPL value and the original return address, as per your example.)

In the case of an interrupt storm:

a) timer interrupt occurs

b) CPU creates new stack frame for the interrupt

c) TimerInterruptHandler() issues EOI

d) TimerInterruptHandler() rearranges the stack frame to allow for EventProcessingTrampoline to be invoked after IRET.  This rearrangement consumes additional stack space.

e) TimerInterruptHandler() returns via IRET to EventProcessingTrampoline.

f) timer interrupt occurs immediately after IRET

g) go to step (b) but with the additional stack space not yet released

I don&apos;t see how the stack underrun is prevented in this scenario?

&gt; That said, I will look at your patch.

Thank you, much appreciated!

Michael

**Comment by tianocore-issues** [comment]
## Comment 20096
Date: 2022-11-24 13:41:03 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

tl;dr: your patch is scary af but seems to be fine. :)  You can skip my answers if you wish and jump to the discussion of your patch.

If I understood your patch correctly, your patch in the nested event case does something similar to what I intended:

      TargetTPL == 0
      ...
      OldTargetTPL = 0
      TargetTPL = TPL_APPLICATION
      RestoreTPLPending &amp;= ~(1 &lt;&lt; TPL_APPLICATION)
      gBS-&gt;RestoreTPL (TPL_APPLICATION);
        EnableInterrupts();
        dispatch a TPL_CALLBACK event
          ... schedule a TPL_NOTIFY event ...
          inner TimerInterruptHandler
            RaiseTPL(TPL_HIGH_LEVEL);  // returns TPL_NOTIFY
            ...
            OldTargetTPL = TPL_APPLICATION
            TargetTPL = TPL_CALLBACK
            RestoreTPL(TPL_CALLBACK);
              ... dispatch TPL_NOTIFY event ...
            DisableInterrupts();
            TargetTPL = TPL_APPLICATION;
            IRET re-enables interrupts
          ... finishes dispatching TPL_CALLBACK event ...
      DisableInterrupts();
      checks again
      TargetTPL = 0;
      IRET

Instead, this is the hand-off case:

      TargetTPL == 0
      ...
      OldTargetTPL = 0
      TargetTPL = TPL_APPLICATION
      RestoreTPLPending &amp;= ~(1 &lt;&lt; TPL_APPLICATION)
      gBS-&gt;RestoreTPL (TPL_APPLICATION);
        EnableInterrupts();
        gEfiCurrentTpl = TPL_APPLICATION;         // the last assignment before returning
      inner TimerInterruptHandler
        OriginalTPL = RaiseTPL(TPL_HIGH_LEVEL);   // OriginalTPL == TargetTPL, so no event was being dispatched
        sees OriginalTPL == TargetTPL
          RestoreTPLPending |= (1 &lt;&lt; TPL_APPLICATION);
        IRET                                                    (1)
      DisableInterrupts();                                      (2)
      loop executes again!
      gBS-&gt;RestoreTPL (TPL_APPLICATION);
        EnableInterrupts();
      DisableInterrupts();
      checks again
      TargetTPL = 0;
      IRET

The good: it&apos;s important that the outer RestoreTPL() was about to return, because it means it was not dispatching any event.  Thus, skipping the inner RestoreTPL() thus will not cause a priority inversion.

The ugly: between (1) and (2) you have interrupts enabled but gEfiCurrentTpl == TPL_HIGH_LEVEL.  In theory this is a serious invariant violation that could mess things up royally, because CoreRaiseTpl(TPL_HIGH_LEVEL) in the interrupt handler will *not* disable interrupts.  But the only thing that can happen is that another interrupts comes, and before the handler is entered the process will helpfully clear IF=0 and restore the invariant.  Pfew.

So that&apos;s why I called it scary stuff, :) it would be nicer if it were possible to modify the flags on the interrupt frame and keep IF=0 after the IRET at (1).  But I don&apos;t think that&apos;s possible, and it seems to work.  Only needs a lot of comments and somebody else going after my thoughts here.

**Comment by tianocore-issues** [comment]
## Comment 20097
Date: 2022-11-24 16:22:44 +0000
From: Michael Brown &lt;<mcb30>&gt;

(In reply to Paolo Bonzini from comment #17)
&gt; The ugly: between (1) and (2) you have interrupts enabled but gEfiCurrentTpl
&gt; == TPL_HIGH_LEVEL.  In theory this is a serious invariant violation that
&gt; could mess things up royally, because CoreRaiseTpl(TPL_HIGH_LEVEL) in the
&gt; interrupt handler will *not* disable interrupts.  But the only thing that
&gt; can happen is that another interrupts comes, and before the handler is
&gt; entered the process will helpfully clear IF=0 and restore the invariant. 
&gt; Pfew.
&gt; 
&gt; So that&apos;s why I called it scary stuff, :) it would be nicer if it were
&gt; possible to modify the flags on the interrupt frame and keep IF=0 after the
&gt; IRET at (1).  But I don&apos;t think that&apos;s possible, and it seems to work.  Only
&gt; needs a lot of comments and somebody else going after my thoughts here.

I totally agree that it would be much if we could modify the inner interrupt stack frame to set IF=0 in the stored EFLAGS to avoid violating the invariant.

From digging through ExceptionHandlerAsm.nasm, it looks as though modifying SystemContext-&gt;Rflags might work for this.  I&apos;ll give it a try.

Michael

**Comment by tianocore-issues** [comment]
## Comment 20098
Date: 2022-11-24 16:55:53 +0000
From: Michael Brown &lt;<mcb30>&gt;

&gt; From digging through ExceptionHandlerAsm.nasm, it looks as though modifying
&gt; SystemContext-&gt;Rflags might work for this.  I&apos;ll give it a try.

It does work.  :)

Working on an updated patch now...

Michael

**Comment by tianocore-issues** [comment]
## Comment 20099
Date: 2022-11-24 18:29:14 +0000
From: Michael Brown &lt;<mcb30>&gt;

Here&apos;s the updated code, which now maintains the invariant that interrupts are always disabled when at TPL_HIGH_LEVEL, and is structured to provide an abstraction that can be shared between LocalApicTimerDxe and 8254TimerDxe:

//
// State shared between all invocations of a nested interrupt handler.
//
typedef struct {
  //
  // Target TPL for the innermost in-progress call to RestoreTPL().
  //
  EFI_TPL  TargetTPL;
  //
  // Bitmap used to by an inner interrupt handler to defer a call to
  // RestoreTPL() to the outer interrupt handler.
  //
  UINTN  RestoreTPLPending;
} NESTED_INTERRUPT_STATE;

/**
  Raise the task priority level to TPL_HIGH_LEVEL from within a nested
  interrupt handler.

  @param  None.

  @return The previous task priority level

**/
EFI_TPL
EFIAPI
NestedInterruptRaiseTPL (
  VOID
  )
{
  EFI_TPL OriginalTPL;

  //
  // Raise TPL and assert that we were called from within an interrupt
  // handler (i.e. with TPL below TPL_HIGH_LEVEL but with interrupts
  // disabled).
  //
  ASSERT (GetInterruptState() == FALSE);
  OriginalTPL = gBS-&gt;RaiseTPL (TPL_HIGH_LEVEL);
  ASSERT (OriginalTPL &lt; TPL_HIGH_LEVEL);

  return OriginalTPL;
}

/**
  Lower the task priority to the previous value from within a nested
  interrupt handler.  To avoid stack exhaustion, this call may (when
  provably safe to do so) defer the actual lowering to be performed by
  an outer interrupt handler.

  @param OriginalTPL           Previous TPL from NestedInterruptRaiseTPL().

  @param SystemContext         A pointer to the system context when the
                               interrupt occurred.

  @param NestedInterruptState  A pointer to the state shared between all
                               invocations of the nested interrupt handler.

**/
VOID
EFIAPI
NestedInterruptRestoreTPL (
  IN EFI_TPL                     OriginalTPL,
  IN OUT EFI_SYSTEM_CONTEXT      SystemContext,
  IN OUT NESTED_INTERRUPT_STATE  *NestedInterruptState
  )
{
  IA32_EFLAGS32  InterruptsEnabledFlag;
  UINTN          *IretFlags;
  UINTN          TPLPendingFlag;
  EFI_TPL        OldTargetTPL;

  //
  // Construct constant for IF flag (without using a C99 initializer).
  //
  InterruptsEnabledFlag.UintN = 0;
  InterruptsEnabledFlag.Bits.IF = 1;

  //
  // Locate flags that will be restored by IRET within system context.
  //
#if defined (MDE_CPU_IA32)
  IretFlags = &amp;SystemContext.SystemContextIa32-&gt;Eflags;
#elif defined (MDE_CPU_X64)
  IretFlags = &amp;SystemContext.SystemContextX64-&gt;Rflags;
#else
  #error &quot;Unsupported CPU&quot;
#endif

  //
  // This is unfortunately messy.  UEFI requires us to support nested
  // interrupts, but provides no way for an interrupt handler to call
  // RestoreTPL() without implicitly re-enabling interrupts.  In a
  // virtual machine, it is possible for a large burst of interrupts
  // to arrive.  We must prevent such a burst from leading to stack
  // underrun, while continuing to allow nested interrupts to occur.
  //
  // Since nested interrupts are permitted, an interrupt handler may
  // be invoked as an inner interrupt handler while the outer
  // interrupt handler is still inside its call to RestoreTPL().
  //
  // If the TPL at which this interrupt occurred is higher than the
  // target TPL for the outer handler&apos;s call to RestoreTPL(), then the
  // outer handler&apos;s call to RestoreTPL() is still dispatching event
  // notifications.  In this circumstance we must call RestoreTPL() to
  // allow more event notifications to be dispatched, since those
  // callbacks may themselves be waiting upon events.  We cannot avoid
  // creating a new stack frame for this inner call to RestoreTPL(),
  // but the total number of such stack frames is intrinsically
  // limited by the number of distinct TPLs.
  //
  // If the TPL at which this interrupt occurred is equal to the
  // target TPL for the outer handler&apos;s call to RestoreTPL(), then the
  // outer handler&apos;s call to RestoreTPL() must have finished
  // dispatching all event notifications, and the interrupt therefore
  // occurred at the point that RestoreTPL() was about to return to
  // the outer handler.  In this circumstance, we must not call
  // RestoreTPL() here since doing so could lead to stack exhaustion.
  // Instead, we return from the ISR at TPL_HIGH_LEVEL with interrupts
  // still disabled and set a flag to trigger the outer handler into
  // performing the deferred call to RestoreTPL().
  //
  ASSERT (OriginalTPL &gt;= NestedInterruptState-&gt;TargetTPL);
  TPLPendingFlag = (UINTN)(1U &lt;&lt; OriginalTPL);
  if (OriginalTPL == NestedInterruptState-&gt;TargetTPL) {
    ASSERT (*IretFlags &amp; InterruptsEnabledFlag.UintN);
    *IretFlags &amp;= ~InterruptsEnabledFlag.UintN;
    NestedInterruptState-&gt;RestoreTPLPending |= TPLPendingFlag;
  } else {
    OldTargetTPL = NestedInterruptState-&gt;TargetTPL;
    NestedInterruptState-&gt;TargetTPL = OriginalTPL;
    do {
      NestedInterruptState-&gt;RestoreTPLPending &amp;= ~TPLPendingFlag;
      gBS-&gt;RestoreTPL (OriginalTPL);
      DisableInterrupts();
    } while (NestedInterruptState-&gt;RestoreTPLPending &amp; TPLPendingFlag);
    NestedInterruptState-&gt;TargetTPL = OldTargetTPL;
  }
}
  
/**
  Interrupt Handler.

  @param InterruptType    The type of interrupt that occurred
  @param SystemContext    A pointer to the system context when the interrupt occurred
**/
VOID
EFIAPI
TimerInterruptHandler (
  IN EFI_EXCEPTION_TYPE  InterruptType,
  IN EFI_SYSTEM_CONTEXT  SystemContext
  )
{
  STATIC NESTED_INTERRUPT_STATE  NestedInterruptState;
  EFI_TPL                        OriginalTPL;

  OriginalTPL = NestedInterruptRaiseTPL();

  SendApicEoi();

  if (mTimerNotifyFunction != NULL) {
    //
    // @bug : This does not handle missed timer interrupts
    //
    mTimerNotifyFunction (mTimerPeriod);
  }

  NestedInterruptRestoreTPL (OriginalTPL, SystemContext, &amp;NestedInterruptState);
}



@Laszlo @Paolo: any feedback would be very welcome!

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20100
Date: 2022-11-24 18:34:12 +0000
From: Paolo Bonzini &lt;<pbonzini>&gt;

Yes, that should work and is as beautiful as it can be I guess.  Would you like someone else to come up with a patch?

**Comment by tianocore-issues** [comment]
## Comment 20101
Date: 2022-11-24 19:50:07 +0000
From: Michael Brown &lt;<mcb30>&gt;

(In reply to Paolo Bonzini from comment #21)
&gt; Yes, that should work and is as beautiful as it can be I guess.  Would you
&gt; like someone else to come up with a patch?

Thank you!

I&apos;m happy to put together the patch series.  Where within the EDK2 tree should NestedInterruptRestoreTPL() and NestedInterruptRaiseTPL() live so that they can be used by both LocalApicTimerDxe and 8254TimerDxe?

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20102
Date: 2022-11-25 02:48:59 +0000
From: @lersek

Hi Michael,

It&apos;s great if the problem can be solved in OvmfPkg, without touching core edk2 code.

I suggest introducing a new library class, and instance, for that. I think you could call it NestedInterruptTplLib.

I&apos;ve tried to locate an existent commit that (a) is minimal and (b) shows, structurally speaking, how to introduce a new library class (= declaration) with a single instance (= implementation).

I&apos;ve come up with commit 77874ceebb11 (&quot;OvmfPkg: add library to track boot option loading/starting on the console&quot;, 2019-02-25); I recommend following the structure of that. Some comments:

patch#1: revert 239b50a86

patch#2:

- OvmfPkg/OvmfPkg.dec: declare the new library class in the [LibraryClasses] section, referencing the new lib class header file

- OvmfPkg/*.dsc: do not modify these files just yet

- OvmfPkg/Include/Library/NestedInterruptTplLib.h: place the new header file here. Be sure to add #include guards, #include the proper dependencies (just so the function prototypes are self-contained).

- OvmfPkg/Library/NestedInterruptTplLib/NestedInterruptTplLib.inf: you can mostly follow &quot;PlatformBmPrintScLib.inf&quot; here. INF_VERSION should be set to the highest version currently documented in the edk2 INF file spec. FILE_GUID should be generated afresh with &quot;uuidgen&quot;. The [Packages] section must contain &quot;MdePkg/MdePkg.dec&quot;, but other than that, keep it minimal -- only add packages whose header files and library classes you actually depend on. [LibraryClasses] should again be minimal -- it generally matches the lib class #includes from &quot;NestedInterruptTplLib.h&quot;. The [Pcd], [Protocols] and [Guids] sections you don&apos;t need I think. [Depex.common.DXE_DRIVER] is required; AFAICT you can just set it to TRUE.

- OvmfPkg/Library/NestedInterruptTplLib/NestedInterruptTplLib.c: this is where the function definitions (= implementation) should go.

patch#3:

- make use of NestedInterruptTplLib in LocalApicTimerDxe and 8254TimerDxe. This means extending their #include directives (C source), and also their [LibraryClasses] sections (INF file). If possible, keep both the #include directives and the [LibraryClasses] sections alphabetically sorted.

- OvmfPkg/*.dsc: locate all those platform DSC files that include either LocalApicTimerDxe or 8254TimerDxe. In these DSC files, resolve the new lib class to the new (sole) lib instance. Because the library instance is going to suit DXE_DRIVER modules only, it&apos;s best to add the lib class resolution to the [LibraryClasses.common.DXE_DRIVER] section of each affected DSC file.

**Comment by tianocore-issues** [comment]
## Comment 20103
Date: 2022-11-25 02:50:42 +0000
From: @lersek

(In reply to Laszlo Ersek from comment #23)

&gt; patch#2:
&gt; 
&gt; - OvmfPkg/OvmfPkg.dec: declare the new library class in the [LibraryClasses]
&gt; section, referencing the new lib class header file

keep this section lexicographically ordered too, if possible, pls

**Comment by tianocore-issues** [comment]
## Comment 20104
Date: 2022-11-25 08:59:52 +0000
From: Michael Brown &lt;<mcb30>&gt;

(In reply to Laszlo Ersek from comment #23)
&gt; I&apos;ve tried to locate an existent commit that (a) is minimal and (b) shows,
&gt; structurally speaking, how to introduce a new library class (= declaration)
&gt; with a single instance (= implementation).
&gt; 
&gt; I&apos;ve come up with commit 77874ceebb11 (&quot;OvmfPkg: add library to track boot
&gt; option loading/starting on the console&quot;, 2019-02-25); I recommend following
&gt; the structure of that.

Thank you so much; that&apos;s incredibly helpful.  I&apos;ll pull something together now following your instructions.

I&apos;m guessing that ArmPkg/Drivers/TimerDxe/TimerDxe.c may need a similar fix, since the underlying stack underrun problem is not x86-specific (or even VM-specific, though it&apos;s much more easily triggerable in a VM).  Would this suggest placing NestedInterruptTplLib outside of OvmfPkg?  And, if so, where ought it to go?

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20105
Date: 2022-11-25 09:57:10 +0000
From: @lersek

Normally if you want to share code (library or driver) between OvmfPkg and ArmVirtPkg, then you put the component (library or driver) under OvmfPkg, and just consume it from ArmVirtPkg. ArmVirtPkg is welcome to consume components from OvmfPkg. The reverse dependency is not allowed.

However, this question is trickier than that. If we do not consider ARM / AARCH64, then the patch set is contained under OvmfPkg, so it only affects virtual hardware. Put differently, the virtual machine platforms (DSC files) under OvmfPkg all use such Timer DXE_DRIVER modules that also exist under OvmfPkg -- in other words, the IA32/X64 VM platforms all use VM-specific timer drivers.

But that&apos;s not the case with ArmVirtPkg. The DSC files there use &quot;ArmPkg/Drivers/TimerDxe/TimerDxe.inf&quot;, but that driver is included by some physical firmware platforms too (in the &quot;edk2-platforms&quot; tree, for example). So you cannot modify ArmPkg/Drivers/TimerDxe as easily as you can the two TimerDxe drivers under OvmfPkg, for two reasons: (a) the changes will affect real ARM / AARCH64 hardware, where the original issue may never happen in practice, (b) ArmPkg content -- where the TimerDxe driver in question lives -- is not allowed to depend on OvmfPkg, for example on a new library under OvmfPkg.

Problem (b) is easy to solve. For such cases, we put components (libraries, drivers, GUID definitions, PCDs etc) under EmbeddedPkg. All of ArmPkg, ArmVirtPkg, and OvmfPkg are permitted to depend on EmbeddedPkg.

Therefore, *if* you intend to modify ArmPkg/Drivers/TimerDxe as well, then the new library should go under EmbeddedPkg. Furthermore, the patch modifying ArmPkg/Drivers/TimerDxe needs to stand alone; separate from the patch that rebases the timer drivers in OvmfPkg to the new library.

**Comment by tianocore-issues** [comment]
## Comment 20106
Date: 2022-11-25 10:18:20 +0000
From: @lersek

(In reply to Laszlo Ersek from comment #26)

&gt; If we do not consider ARM / AARCH64, then the patch set is contained
&gt; under OvmfPkg, so it only affects virtual hardware. Put differently,
&gt; the virtual machine platforms (DSC files) under OvmfPkg all use such
&gt; Timer DXE_DRIVER modules that also exist under OvmfPkg -- in other
&gt; words, the IA32/X64 VM platforms all use VM-specific timer drivers.

Small correction there -- the bhyve platform firmware
&quot;OvmfPkg/Bhyve/BhyveX64.dsc&quot; does consume a phys hardware oriented timer
driver: &quot;PcAtChipsetPkg/HpetTimerDxe/HpetTimerDxe.inf&quot;. I don&apos;t know if
that driver is affected too, but if it is, and you intend to modify it,
then the new library belongs under PcAtChipsetPkg (if you steer clear of
ARM/AARCH64), or else MdeModulePkg (if you want to consume the library
from both ARM/AARCH64 and IA32/X64).

... For completeness (?), I think edk2 contains the following timer
drivers (i.e., DXE_DRIVER modules producing
&quot;gEfiTimerArchProtocolGuid&quot;):

  ArmPkg/Drivers/TimerDxe/TimerDxe.inf
  EmulatorPkg/TimerDxe/Timer.inf
  OvmfPkg/8254TimerDxe/8254Timer.inf
  OvmfPkg/LocalApicTimerDxe/LocalApicTimerDxe.inf
  PcAtChipsetPkg/HpetTimerDxe/HpetTimerDxe.inf

I&apos;d suggest modifying only those in which you can reproduce the symptom
(and test the fix). It&apos;s always possible to move a library to a
different package, for broader consumption.

**Comment by tianocore-issues** [comment]
## Comment 20107
Date: 2022-11-25 10:32:17 +0000
From: @lersek

So, to provide a rudimentary map:

  Timer drivers modified in addition
  to those under OvmfPkg
  -----------------------------------
  ArmPkg  EmulatorPkg  PcAtChipsetPkg  Library to be placed under package
  ------  -----------  --------------  -----------------------------------
       0            0               0  OvmfPkg
       0            0               1  PcAtChipsetPkg
       0            1               0  EmbeddedPkg
       0            1               1  MdeModulePkg
       1            0               0  EmbeddedPkg
       1            0               1  MdeModulePkg
       1            1               0  EmbeddedPkg
       1            1               1  MdeModulePkg

**Comment by tianocore-issues** [comment]
## Comment 20108
Date: 2022-11-25 11:40:31 +0000
From: Michael Brown &lt;<mcb30>&gt;

(In reply to Laszlo Ersek from comment #27)
&gt; I&apos;d suggest modifying only those in which you can reproduce the symptom
&gt; (and test the fix). It&apos;s always possible to move a library to a
&gt; different package, for broader consumption.

Thank you once again for such detailed help!

I&apos;ve created a PR to fix only OvmfPkg/LocalApicTimerDxe and OvmgPkg/8254TimerDxe for now, on the basis that those are the two drivers where nested interrupts are currently broken.  Once it gets merged, I&apos;ll talk to some people at ARM to find out if there&apos;s any appetite for fixing the potential stack underrun in ArmPkg/TimerDxe.

The PR is at https://github.com/tianocore/edk2/pull/3662 and is currently awaiting CI results.

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20111
Date: 2022-11-28 07:10:03 +0000
From: @lersek

The logic boggles my mind. In trying to explain it to myself, I&apos;ve

(a) renamed some variables -- I&apos;ve done this because &quot;target&quot;, &quot;original&quot;, and &quot;oldtarget&quot; were way too obscure for me

(b) replaced the

  if (...) { ... } else { ... }

in the interrupt handler with

  if (...) { ...; return; }
  ...

because I think the nesting of the second branch there -- where we loop -- is really detrimental to understanding; it suggests that something else is coming after the first banch. My understanding is helped by seeing that the first branch terminates the function right there.

(c) I&apos;ve rearranged the ASSERT a bit, just to highlight the TPL ordering / relationships

(d) when a bit-and is evaluated in a boolean context, edk2 requires an explicit comparison against zero; so I&apos;ve updated the controlling expression in the do { ... } while (...).

Now, the names are inane, sorry about them; again this is just for explaining the logic to myself.

VOID
EFIAPI
NestedInterruptRestoreTPL (
  IN EFI_TPL                     TplToRestoreNow,
  IN OUT EFI_SYSTEM_CONTEXT      SystemContext,
  IN OUT NESTED_INTERRUPT_STATE  *NestedInterruptState
  )
{
  UINTN    TplToRestoreNowBit;
  EFI_TPL  LowerTplWhoseRestorationHasBeenInterrupted;

  TplToRestoreNowBit = (UINTN)(1U &lt;&lt; TplToRestoreNow);

  if (TplToRestoreNow ==
      NestedInterruptState-&gt;HighestTplBeingRestored) {
    DisableInterruptsOnIret (SystemContext);
    NestedInterruptState-&gt;TplsToRestoreLaterMask |= TplToRestoreNowBit;
    return;
  }

  ASSERT (TplToRestoreNow &gt;
          NestedInterruptState-&gt;HighestTplBeingRestored);

  LowerTplWhoseRestorationHasBeenInterrupted =
    NestedInterruptState-&gt;HighestTplBeingRestored;
  NestedInterruptState-&gt;HighestTplBeingRestored = TplToRestoreNow;

  do {
    NestedInterruptState-&gt;TplsToRestoreLaterMask &amp;= TplToRestoreNowBit;
    gBS-&gt;RestoreTPL (TplToRestoreNow);
    DisableInterrupts ();
  } while ((NestedInterruptState-&gt;TplsToRestoreLaterMask &amp;
            TplToRestoreNowBit) != 0);

  NestedInterruptState-&gt;HighestTplBeingRestored =
    LowerTplWhoseRestorationHasBeenInterrupted;
}


Now, I think the patch needs some more comments / a bit longer commit message than currently visible at PR#3662, with branch nestedinterrupttpllib @ 5c1718b26a95, so that mere mortals like myself can understand it:

(1) the DisableInterruptsOnIret() and DisableInterrupts() calls need more explanation.

(1.1) My understanding of DisableInterrupts() is that it protects the &quot;TplsToRestoreLaterMask&quot; manipulation just gBS-&gt;RestoreTPL() returns, with interrupts enabled. We don&apos;t want the &quot;short path&quot; to queue another restoration right after we&apos;ve determined there is nothing queued. That seems reasonable, but it deserves a comment (not to mention if there is a different reason, or a different reason *too*, for the DisableInterrupts() call).

(1.2) The DisableInterruptsOnIret() call is somewhat explained in comments 17 and 18; that reasoning should be boiled down to a code comment. Namely (IIUC), in the &quot;short branch&quot;, at the time of calling DisableInterruptsOnIret(), the current TPL *from which* we&apos;re about to restore, is &quot;HIGH&quot;, and we must not exit the interrupt handler with the current TPL being HIGH *and* interrupts being enabled.

Perhaps this could be expressed (*in addition* to the comment) with

  ASSERT (gBS-&gt;RaiseTPL (TPL_HIGH_LEVEL) == TPL_HIGH_LEVEL);

(2) For a human, following recursive entries is easy, but following recursive *returns* is darn hard. (Cf. the COMEFROM statement.) By my count, the gBS-&gt;RestoreTPL() call in the function may return from three separate paths:

- from the short branch, after which we&apos;re going to loop
- from a genuinely nested TPL restoration (dispatched handlers justifiedly interrupted)
- with handlers dispatched or not, but nothing interrupted

I could be wrong about these possibilities very easily, so I suggest to spell them out in the code.

Thanks.

**Comment by tianocore-issues** [comment]
## Comment 20112
Date: 2022-11-28 07:12:53 +0000
From: @lersek

(In reply to Laszlo Ersek from comment #30)

&gt; (2) For a human, following recursive entries is easy, but following
&gt; recursive *returns* is darn hard. (Cf. the COMEFROM statement.) By my count,
&gt; the gBS-&gt;RestoreTPL() call in the function may return from three separate
&gt; paths:
&gt; 
&gt; - from the short branch, after which we&apos;re going to loop
&gt; - from a genuinely nested TPL restoration (dispatched handlers justifiedly
&gt; interrupted)
&gt; - with handlers dispatched or not, but nothing interrupted
&gt; 
&gt; I could be wrong about these possibilities very easily, so I suggest to
&gt; spell them out in the code.

Sorry, so that was (2.1); (2.2) would be:

- please include some example call trees, similar to Paolo&apos;s from comment 17, in the commit message. Thinking about the recursive returns makes me feel extremely dumb, like my gourd is filled with sand. :)

**Comment by tianocore-issues** [comment]
## Comment 20185
Date: 2022-12-05 11:27:18 +0000
From: Michael Brown &lt;<mcb30>&gt;

@Laszlo: Thank you for the in-depth review and feedback.  I&apos;ve updated the code to follow a structure closer to that which you suggested.

In the process of doing so, I&apos;ve proved to myself that we don&apos;t need the per-TPL bitmask, and replaced it with a single boolean flag.

I&apos;ve expanded on the existing comments, and added &quot;DEFERRAL INVOCATION POINT&quot; and &quot;DEFERRAL RETURN POINT&quot; comments to explain the control flow that takes place in the complicated case (when we return from an inner handler without calling RestoreTPL()).

I haven&apos;t added an ASSERT(gBS-&gt;RaiseTPL(TPL_HIGH_LEVEL) == TPL_HIGH_LEVEL) because it feels wrong to me to have an assertion that looks as though it has a significant side-effect, and I can&apos;t find any other examples of EDK2 assertions that call boot services methods.  I&apos;m happy to defer to your judgement if you would still prefer this to be added.

I&apos;ve added sample call trees to the commit message as requested.

Thank you again, and please let me know your thoughts on the updated PR!  :)

**Comment by tianocore-issues** [comment]
## Comment 20192
Date: 2022-12-06 03:57:29 +0000
From: @lersek

Hi Michael,

many thanks for addressing my comments; I think your point about not adding the ASSERT() is valid / justified. I think the patch set should be posted upstream now. I might be able to read through the patches at that point again (not sure); unfortunately right now I must content myself with your comment 32 here, as I&apos;m so badly out of time that it&apos;s inexplicable even to myself :/ I don&apos;t want you to wait for me, so please feel free to post the patches upstream.

Laszlo

**Comment by tianocore-issues** [comment]
## Comment 20200
Date: 2022-12-09 05:27:49 +0000
From: Michael Brown &lt;<mcb30>&gt;

Finally managed to fix my outbound mail configuration and get it sent upstream: https://edk2.groups.io/g/devel/message/97173

Thanks,

Michael

**Comment by tianocore-issues** [comment]
## Comment 20281
Date: 2022-12-23 12:00:59 +0000
From: Michael Brown &lt;<mcb30>&gt;

Merged upstream via https://github.com/tianocore/edk2/pull/3829
